### PR TITLE
Life cycle

### PR DESCRIPTION
**Pull request by katrina433**
Added a database that stores values of animals in each class. 
Initialize herd by drawing animals from the database. 
Changed animal constructors to either initialize or assign values based on the input. 
Limited inputs to be within two standard deviations from the average.

### CODE REVIEW

**Comment by militsasotirova** [comment]
Overall, it looks good! I like how you've organized the animal_initialization & simplified some of the existing code. 
A few suggestions: In your get_calves(), get_heiferIs(), get_heiferIIs(), etc. methods, it might be good to check whether 'num' is greater than the amount of rows in the respective tables (prevent user error), and if it is you could raise a warning and set 'num' to be the actual amount of rows in that table (sort of like a ceiling). Also, you could use a LIMIT clause in your SELECT query so that you don't have to have that for loop in each of those methods - you would get back from the DB exactly as many rows as you need.
Could you add a bit more detailed comments for those methods?

**Review by militsasotirova — APPROVED** [review]

File: RUFAS/routines/animal/life_cycle/calf.py
Comment (wmaxdonovan) [inline RUFAS/routines/animal/life_cycle/calf.py]: one thought here is that you may want to prematurely pull changes from the phosphorus tracking branch (assuming that the plan is for phosphorus tracking to be merged first) because I believe @militsasotirova changed these from protected to public class variables.

File: RUFAS/routines/animal/life_cycle/calf.py
Comment (wmaxdonovan) [inline RUFAS/routines/animal/life_cycle/calf.py]: I think this while statement breaks the 120c line limit. I'm more concerned with readability than outdated compiler errors though, so I will defer to you on the best way to group these statements.

File: RUFAS/routines/animal/life_cycle/cow.py
Comment (wmaxdonovan) [inline RUFAS/routines/animal/life_cycle/cow.py]: Same thing as above comment. Maybe consider breaking at the or.

File: RUFAS/routines/animal/life_cycle/heiferI.py
Comment (wmaxdonovan) [inline RUFAS/routines/animal/life_cycle/heiferI.py]: line length flag

Comment (wmaxdonovan) [reply RUFAS/routines/animal/life_cycle/heiferI.py]: Also, feel free to resolve my comments whenever you've determined a solution. They aren't providing a ton of value beyond flagging

File: RUFAS/routines/animal/life_cycle/heiferII.py
Comment (wmaxdonovan) [inline RUFAS/routines/animal/life_cycle/heiferII.py:157]: Same thing with these variables– I think @militsasotirova made these public in phosphorus cycling

File: RUFAS/routines/animal/life_cycle/heiferII.py
Comment (wmaxdonovan) [inline RUFAS/routines/animal/life_cycle/heiferII.py]: line length flag

File: RUFAS/routines/animal/life_cycle/heiferII.py
Comment (wmaxdonovan) [inline RUFAS/routines/animal/life_cycle/heiferII.py]: line length flag

File: RUFAS/routines/animal/life_cycle/heiferII.py
Comment (wmaxdonovan) [inline RUFAS/routines/animal/life_cycle/heiferII.py]: line length flag

File: RUFAS/routines/animal/life_cycle/heiferIII.py
Comment (wmaxdonovan) [inline RUFAS/routines/animal/life_cycle/heiferIII.py]: line length flag

**Comment by wmaxdonovan** [comment]
Hey Katrina, I like what you've done! The line length stuff is a readability thing. The only change that I would like to see before merging is the protected variable classification. Militsa changed this in phosphorus tracking, so we can either wait for that pull request to be merged (and then pull from master up to this branch and make sure there are no left over protected variables), or prematurely pull from phosphorus tracking just to get a head start so this will be ready to go when phosphorus tracking is merged.

**Comment by wmaxdonovan** [comment]
Oh! And don't forget to add yourself as an author on the files you edit in a significant way.

**Review by wmaxdonovan — APPROVED** [review]
Looks good Katrina! Is this dependent on the phosphorus tracking branch or is it being merged prior?

**Comment by katrina433** [comment]
It is not merged with the P tracking branch. I just changed the variable names on this branch.

**Review by wmaxdonovan — APPROVED** [review]
Added a very minor commit to fix a very minor error resulting from split() on a string with differential capitalization. Other than that I like these changes and think the code looks good.

File: RUFAS/routines/animal/life_cycle/animal_initialization.py
Comment (tlh96) [inline RUFAS/routines/animal/life_cycle/animal_initialization.py:188]: Are these user inputs?

File: RUFAS/routines/animal/life_cycle/animal_initialization.py
Comment (tlh96) [inline RUFAS/routines/animal/life_cycle/animal_initialization.py]: p_gest_calf is > 0 for gestation length greater than 190 days. Should that be included as a value for heifer II, or does it not matter since we aren't purchasing heifer II?

File: RUFAS/routines/animal/life_cycle/heiferI.py
Comment (tlh96) [inline RUFAS/routines/animal/life_cycle/heiferI.py]: Minor suggestion: The description would be clearer if you explicitly stated the inputs as in calf. Some users may not view all the files.

	Input:
			args.id: id of the cow
			args.breed: breed of the cow
			args.birth_date: the date of the simulation when the calf was born
			args.daysBorn: age of the animal
			(optional: include the following to assign cow information) 
			args.birth_weight: the birth weight of the cow
			args.body_weight: current body weight of the cow
			args.wean_weight: the wean weight of the cow
			args.mature_body_weight: the mature body weight of the cow
			args.events: events of the cow

File: RUFAS/routines/animal/life_cycle/animal_initialization.py
Comment (katrina433) [reply RUFAS/routines/animal/life_cycle/animal_initialization.py:188]: No, this was also hardcoded in the original initialize_herd() method in life_cycle.py

File: RUFAS/routines/animal/life_cycle/animal_initialization.py
Comment (katrina433) [reply RUFAS/routines/animal/life_cycle/animal_initialization.py]: When I initialized animals for the database, I did not call any P tracking methods so none of the P tracking values are being updated, including p_gest_calf. I just realized this value is 0 in the heiferIII and cow tables as well. Should I update and store p_gest_calf in the database or should I initialize animals with 0 so it will only update once the animals enter the herd?

File: RUFAS/routines/animal/life_cycle/animal_initialization.py
Comment (manfei-L) [reply RUFAS/routines/animal/life_cycle/animal_initialization.py:188]: this will be reorganized in the life-cycle-output branch

File: RUFAS/routines/animal/life_cycle/heiferI.py
Comment (manfei-L) [reply RUFAS/routines/animal/life_cycle/heiferI.py]: Thanks for the reminder of formatting those comments. They are going to be improved along with the removal of some unwanted calculations in the pull request after we have the sensitivity analysis branch.

File: RUFAS/routines/animal/life_cycle/animal_initialization.py
Comment (tlh96) [reply RUFAS/routines/animal/life_cycle/animal_initialization.py]: Oh, that's a good question. We originally thought P initial values could be a percentage of bodyweight for purchased animals. Now, I'm not sure that will work with the replacement herd. I'll mark as an issue for later.

From what I can tell, you added P tracking with the new commit. That should work for now.

**Review by tlh96 — APPROVED** [review]
Overall, the animal initialization is easy to follow and a steady state is reached much faster than previous versions. Some minor commenting suggestions are listed. Great work!

**Pull request merged**
